Removed redundant ChangingDatabasesByForksTest

grzesiek2010 opened the pull request:
#### What has been done to verify that this works as intended?
Just reviewed implemented changes.

#### Why is this the best possible solution? Were any other approaches considered?
`ChangingDatabasesByForksTest` is sometimes failing on firebase as you have probably noticed. Since we migrated to scoped storage and we no longer share the same directory (and databases) with forks this test is redundant because other apps can't use and edit our databases.

#### How does this change affect users? Describe intentional changes to behavior and behavior that could have accidentally been affected by code changes. In other words, what are the regression risks?
It doesn't require testing.

#### Do we need any specific form for testing your changes? If so, please attach one.
No.

#### Does this change require updates to documentation? If so, please file an issue [here]( https://github.com/getodk/docs/issues/new) and include the link below.
No.

#### Before submitting this PR, please make sure you have:
- [x] run `./gradlew checkAll` and confirmed all checks still pass OR confirm CircleCI build passes and run `./gradlew connectedDebugAndroidTest` locally.
- [x] verified that any code or assets from external sources are properly credited in comments and/or in the [about file](https://github.com/getodk/collect/blob/master/collect_app/src/main/assets/open_source_licenses.html).
- [x] verified that any new UI elements use theme colors. [UI Components Style guidelines](https://github.com/getodk/collect/blob/master/CONTRIBUTING.md#ui-components-style-guidelines)

grzesiek2010 committed:
Removed redundant ChangingDatabasesByForksTest

grzesiek2010 commented:
@lognaturel 
we could also get rid of:
https://github.com/getodk/collect/blob/022d8dcabf3940cc5978fb308b66d6955d59bea9/collect_app/src/main/java/org/odk/collect/android/database/FormsDatabaseHelper.java#L94

or maybe you would like to keep it just in case (i don't know what case exactly :smile:)?

lognaturel commented:
> we could also get rid of

I think we should as well as the corresponding `instance` method. But how about we do that at the beginning of v1.30 so it has a bit to get exercised? I think it's right that the only way the databases would be out of date would be if a fork touched them but it feels like there's a tiny bit of risk there.

grzesiek2010 committed:
Removed the extra check if form db should be upgraded

grzesiek2010 committed:
Removed the extra check if instances db should be upgraded

grzesiek2010 committed:
Code improvements

grzesiek2010 committed:
Factored out InstanceDatabaseMigrator to follow the same structre of classes we have in acase of forms db

grzesiek2010 committed:
Implemented interface for database migrators

grzesiek2010 committed:
Moved constants to DatabaseConstants class

grzesiek2010 committed:
Pass StoragePathProvider to database helpers

grzesiek2010 committed:
Added missing log

grzesiek2010 commented:
> I think we should as well as the corresponding instance method. But how about we do that at the beginning of v1.30 so it has a bit to get exercised?

I agree let's wait with this pr. I removed those redundant checks and added some refactoring as well.

seadowg commented on collect_app/src/androidTest/java/org/odk/collect/android/instrumented/database/helpers/InstancesDatabaseHelperTest.java:
```diff
@@ -29,7 +31,7 @@
 @RunWith(Parameterized.class)
 @Ignore("`Parameterized` causes problems for Firebase sharding. Probably need to replace this at JUnit level")
 public class InstancesDatabaseHelperTest extends SqlLiteHelperTest {
```
It looks like these tests all fail. Should we rework or delete (if we don't need them) as part of this?

grzesiek2010 replied:
I think it would be good to have such tests that confirm that upgrading/downgrading databases work well so I would keep it.

seadowg replied:
Totally agree that we want that coverage. With forms, it seems we have that covered in `FormDatabaseMigratorTest` so we could add a similar test for instances and remove this one (so replace the instrumented test with a Robolectric test). That said this PR definitely offers a bunch of good stuff, so we could just create an issue for that?

grzesiek2010 replied:
Yeah I agree.

seadowg reviewed (changes requested):
Looks great! Just a question really about what to do with the ignored tests.

seadowg commented:
@getodk-bot label "needs testing"

I think it'd be good to check upgrading and downgrading still as our core testing around that (for instances) is ignored at the moment.

mmarciniak90 commented:
**Tested with success**

Verified on Android versions: 7.0, 8.1, 10.0

Verified cases:
- upgrade from v1.27 to version with PR
- upgrade from v1.28 to version with PR
- downgrade from version with PR to v1.27 - v1.27 refers to /odk directory and data from 1.29 are not available. I don't think this is a big deal because downgrading is not possible for the end user.
- downgrade from version with PR to v1.28
- settings are not restored to default
- bunch of files like forms, instances, layers available on device during each upgrade/downgrade

Manually generated apks were used to test because downgrade is not possible for released versions and upgrade is possible only if apks are signed by the same key

kkrawczyk123 commented:
Verified on Androids: 5.1, 7.1 and 9.0.

@getodk-bot unlabel "needs testing"
@getodk-bot label "behavior verified"
